src/district/server/smart_contracts.cljs

```diff
@@ -115,132 +107,134 @@
 
 
-(defn- handle-contract-call
-  ([method tx-hash]
-   (handle-contract-call method tx-hash false))
-  ([method tx-hash print-gas-usage?]
-   (let [{:keys [:gas-used :block-number :status]} (web3-eth/get-transaction-receipt @web3 tx-hash)]
-     (when (and gas-used block-number)
-       (when print-gas-usage?
-         (println method (.toLocaleString gas-used) (if (zero? status) "failed" "")))
-       gas-used))))
-
-
 (defn- instance-from-arg [contract]
   (cond
     (keyword? contract) (instance contract)
     (sequential? contract) (instance (first contract) (second contract))
     :else contract))
 
-
-(defn contract-call* [contract-key method & args]
-  "contract-key parameter can be one of:
-   - keyword :some-contract
-   - tuple of keyword and address [:some-contract 0x1234...]
-   - instance SomeContract"
-  (let [deasync? (fn? (last args))
-        callback (when deasync? (last args))
-        args (if deasync? (drop-last args) args)
-        contract (instance-from-arg contract-key)
-        last-arg (last args)
-        args (if (and (map? last-arg)
-                      (not (:from last-arg)))
-               (concat (butlast args) [(merge last-arg {:from (first (web3-eth/accounts @web3))})])
-               args)
-        result (apply web3-eth/contract-call contract method args)
-        filter-id (atom nil)]
-    (if-not (fn? callback)
-      (do
-        (when (and (:print-gas-usage? @smart-contracts)
-                   (map? (last args))
-                   (string? result))
-          (handle-contract-call method result))
-        result)
-      (reset! filter-id
-              (web3-eth/filter
-               @web3
-               "latest"
-               (fn [err]
-                 (when err
-                   (callback err))
-                 (try
-                   (when (and (string? result)
-                              (map? (last args)))
-                     (loop [gas-used (handle-contract-call method result)]
-                       (when-not gas-used
-                         (recur (handle-contract-call method result (:print-gas-usage? @smart-contracts))))))
-                   (web3-eth/stop-watching! @filter-id)
-                   (callback nil result)
-                   (catch js/Error err
-                     (callback err)))))))))
-
-(def contract-call-deasynced (deasync contract-call*))
-
-(defn contract-call [& args]
-  (if (:auto-mining? @smart-contracts)
-    (apply contract-call* args)
-    (apply contract-call-deasynced args)))
+;; TODO : add alts for definite timeout (or # of retries)
+(defn- wait-for-tx-receipt*
+  "callback is a nodejs style callback i.e. (fn [error data] ...)"
+  [tx-hash callback]
+  (web3-eth/get-transaction-receipt @web3 tx-hash (fn [error receipt]
+                                                    (if error
+                                                      (callback error nil)
+                                                      (go
```

Comment:
I think is cleaner here with js/setTimeout instead of go block and timeout channel

Reply:
done.